File: geekbang/algorithm-models/4.6-二叉搜索树后继模板/inorderSuccessor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        return findSucc(root, p.val);
    }


    private TreeNode findSucc(TreeNode root, int val) {
        TreeNode curr = root;
        TreeNode ans = null;
        while (curr != null) {
            if (curr.val > val) { // case2：当后继存在于经过的点中（找到一个>val的最小点）
                // 含义：ans=min(ans, curr.val);
                if (ans == null || ans.val > curr.val)
                    ans = curr;
            }
            if (curr.val == val) {
                if (curr.right != null) { // case1：检索到val且右子树存在，右子树一路向左
                    curr = curr.right;
                    while (curr.left != null) curr = curr.left;
                    return curr;
                }
                break;
            }
            if (val < curr.val) curr = curr.left;
            else curr = curr.right;
        }
        return ans;
    }


}

File: geekbang/algorithm-models/4.6-二叉搜索树后继模板/inorderSuccessor.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderSuccessor(root *TreeNode, p *TreeNode) *TreeNode {
	return findSucc(root, p.Val)
}

func findSucc(root *TreeNode, val int) *TreeNode {
	var ans *TreeNode
	for root != nil {
		if val == root.Val {
			if root.Right != nil {
				p := root.Right
				for p.Left != nil {
					p = p.Left
				}
				return p
			}
			break
		}
		if root.Val > val && (ans == nil || ans.Val > root.Val) {
			ans = root
		}
		if val < root.Val {
			root = root.Left
		} else {
			root = root.Right
		}
	}
	return ans
}
